fdb-record-layer-lucene/src/main/java/org/apache/lucene/codecs/lucene84/LuceneOptimizedPostingsReader.java

```diff
@@ -66,55 +67,66 @@ public final class LuceneOptimizedPostingsReader extends PostingsReaderBase {
-                if (state.fieldInfos.hasPayloads() || state.fieldInfos.hasOffsets()) {
-                    String payName = IndexFileNames.segmentFileName(state.segmentInfo.name, state.segmentSuffix, Lucene84PostingsFormat.PAY_EXTENSION);
-                    payIn = state.directory.openInput(payName, state.context);
+        docIn = Suppliers.memoize(() -> {
+            try {
+                docInInitialized = true;
+                return state.directory.openInput(docName, state.context);
+            } catch (IOException ioe) {
+                throw new UncheckedIOException(ioe);
+            }
+        });
+        version = Suppliers.memoize(() -> {
+            try {
+                return CodecUtil.checkIndexHeader(docIn.get(), DOC_CODEC, VERSION_START, VERSION_CURRENT, state.segmentInfo.getId(), state.segmentSuffix);
+            } catch (IOException ioe) {
+                throw new UncheckedIOException(ioe);
+            }
+        });
+
+        if (state.fieldInfos.hasProx()) {
+            posIn = Suppliers.memoize(() -> {
+                try {
+                    posInInitialized = true;
+                    String proxName = IndexFileNames.segmentFileName(state.segmentInfo.name, state.segmentSuffix, Lucene84PostingsFormat.POS_EXTENSION);
+                    return state.directory.openInput(proxName, state.context);
+                } catch (IOException ioe) {
+                    throw new UncheckedIOException(ioe);
                 }
-            }
-            // The headers are checked last so all 3 openInput calls can begin their fetch of the data concurrently
-            version = CodecUtil.checkIndexHeader(docIn, DOC_CODEC, VERSION_START, VERSION_CURRENT, state.segmentInfo.getId(), state.segmentSuffix);
-            if (posIn != null) {
-                CodecUtil.checkIndexHeader(posIn, POS_CODEC, version, version, state.segmentInfo.getId(), state.segmentSuffix);
-            }
-            if (payIn != null) {
-                CodecUtil.checkIndexHeader(payIn, PAY_CODEC, version, version, state.segmentInfo.getId(), state.segmentSuffix);
-            }
-            this.docIn = docIn;
-            this.posIn = posIn;
-            this.payIn = payIn;
-            success = true;
-        } finally {
-            if (!success) {
-                IOUtils.closeWhileHandlingException(docIn, posIn, payIn);
+            });
+
+            if (state.fieldInfos.hasPayloads() || state.fieldInfos.hasOffsets()) {
+                payIn = Suppliers.memoize(() -> {
+                    try {
+                        payInInitialized = true;
+                        String payName = IndexFileNames.segmentFileName(state.segmentInfo.name, state.segmentSuffix, Lucene84PostingsFormat.PAY_EXTENSION);
+                        return state.directory.openInput(payName, state.context);
+                    } catch (IOException ioe) {
+                        throw new UncheckedIOException(ioe);
+                    }
+                });
```

Comment:
Are the calls to `CodecUtil.checkIndexHeader` omitted on purpose?

Reply:
yes...

Reply:
No reason to do this over and over.